src/env.rs

```diff
@@ -0,0 +1,296 @@
+// Copyright 2019 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! An environment which is passed downward into the widget tree.
+
+use std::collections::HashMap;
+use std::fmt::{Debug, Formatter};
+use std::marker::PhantomData;
+use std::ops::Deref;
+use std::sync::Arc;
+
+use crate::kurbo::{Point, Rect, Size};
+use crate::piet::Color;
+
+use crate::Data;
+
+/// An environment passed down through all widget traversals.
+///
+/// All widget methods have access to an environment, and it is passed
+/// downwards during traversals.
+///
+/// A widget can retrieve theme parameters (colors, dimensions, etc.). In
+/// addition, it can pass custom data down to all descendants. An important
+/// example of the latter is setting a value for enabled/disabled status
+/// so that an entire subtree can be disabled ("grayed out") with one
+/// setting.
+#[derive(Clone, Default)]
+pub struct Env(Arc<EnvImpl>);
+
+#[derive(Clone, Default)]
+struct EnvImpl {
+    map: HashMap<String, Value>,
+}
+
+/// A typed key.
+///
+/// This lets you retrieve values of a given type. The parameter
+/// implements [`ValueType`]. For "expensive" types, this is a reference,
+/// so the type for a string is `Key<&str>`.
+///
+/// [`ValueType`]: trait.ValueType.html
+pub struct Key<T> {
+    key: &'static str,
+    value_type: PhantomData<T>,
+}
+
+// we could do some serious deriving here: the set of types that can be stored
+// could be defined per-app
+// Also consider Box<Any> (though this would also impact debug).
+/// A dynamic type representing all values that can be stored in an environment.
+#[derive(Clone)]
+pub enum Value {
+    Point(Point),
+    Size(Size),
+    Rect(Rect),
+    Color(Color),
```

Comment:
How _would_ we want to include gradient? An option would be to use the `PaintBrush` type, here, or we could have distinct members for `LinearGradient` and `RadialGradient`.

Reply:
When we discussed this on Zulip, consensus was that it should individual variants. But that was before PaintBrush actually landed. I personally like the idea of including PaintBrush, as then we can switch between gradient and solid themes, and the drawing code doesn't have to care.

Reply:
I think that works. One thing I'm realizing is I think `PaintBrush` doesn't derive `PartialEq` (checked, and it actually doesn't derive anything 🤦‍♂)

Reply:
Hmm. I'm running into a problem here, which is that `PaintBrush` isn't `Clone`. That said, I'm thinking the owned type maybe should be `Arc<PaintBrush>`, which of course is `Clone` whether the inner type is or not. I'm going to have to play with types a bit to see how this works, so what I think I'll do it push a commit with all other feedback addressed.